providers/class-two-factor-totp.php

```diff
@@ -136,9 +152,40 @@ public function user_two_factor_options( $user ) {
 			<p>
 				<?php esc_html_e( 'Please scan the QR code or manually enter the key, then enter an authentication code from your app in order to complete setup.', 'two-factor' ); ?>
 			</p>
-			<p>
-				<img src="<?php echo esc_url( $this->get_google_qr_code( $totp_title, $key, $site_name ) ); ?>" id="two-factor-totp-qrcode" />
-			</p>
+			<p id="two-factor-qr-placeholder"></p>
+
+			<style>
+				#two-factor-qr-placeholder {
+					/* The size of the image will change based on the length of the URL inside it. */
+					min-width: 205px;
+					min-height: 205px;
+				}
+			</style>
+
+			<script>
+				window.addEventListener( 'DOMContentLoaded', function( event ) {
+					/*
+					 * 0 = Automatically select the version, to avoid going over the limit of URL
+					 *     length.
+					 * L = Least amount of error correction, because it's not needed when scanning
+					 *     on a monitor, and it lowers the image size.
+					 */
+					var qr     = qrcode( 0, 'L' );
+					var secret = '<?php echo esc_js( $key ); ?>';
+					var label  = encodeURI( '<?php echo esc_js( $totp_title ); ?>' );
+					var issuer = encodeURI( '<?php echo esc_js( $site_name ); ?>' );
+					var url    = 'otpauth://totp/' + label + '?secret=' + secret + '&issuer=' + issuer;
+
+					qr.addData( url );
+					qr.make();
+
+					// ⚠️ Intentionally using GIF here instead of SVG, for security. The benefits
+					// of SVG in this situation are minimal, but it would introduce the possibility
+					// of malicious JS being injected into the SVG and causing XSS.
+					document.getElementById( 'two-factor-qr-placeholder' ).innerHTML = qr.createImgTag( 5 );
+				} );
```

Comment:
IMHO, this is overly cautious code, where it could be much simpler and likely more extendible.

In my testing, I felt the GIF images were slightly blurry, as shown here by GIF and SVG side by side, one is just clearly crisper to my eyes.
<img width="469" alt="Screen Shot 2022-11-08 at 4 21 28 pm" src="https://user-images.githubusercontent.com/767313/200490028-265b43c0-b78a-4cb6-b083-e0ec284d9171.png">

```suggestion
				<?php
				// TODO: Move into the above PHP block after $totp_title.
				$totp_args = array(
					'secret' => rawurlencode( $key ),
					'issuer' => rawurlencode( $site_name )
				);
				$totp_args = apply_filters( 'two_factor_totp_parameters', $totp_args, $user );
	
				$totp_url = 'otpauth://totp/' . rawurlencode( $totp_title );
				$totp_url = add_query_arg( $totp_args, $totp_url );
				$totp_url = apply_filters( 'two_factor_totp_url', $totp_url, $user );
				?>
				window.addEventListener( 'DOMContentLoaded', function( event ) {
					/*
					 * 0 = Automatically select the version, to avoid going over the limit of URL
					 *     length.
					 * L = Least amount of error correction, because it's not needed when scanning
					 *     on a monitor, and it lowers the image size.
					 */
					var qr = qrcode( 0, 'L' );

					qr.addData( <?php echo wp_json_encode( $totp_url ); ?> );
					qr.make();

					document.getElementById( 'two-factor-qr-placeholder' ).innerHTML = qr.createSvgTag( 5 );
				} );
```

I did consider if it was worth moving it into an inline object, `{ type: "0", errorCorrectionLevel: "L", url: ".." }`, but that doesn't gain much, and as you say, I don't think anyone would really want to customise the QR type or errorCorrection level (L = 7%).

Reply:
Huh, those two look identical to me, maybe I'm just old 😆 🤷🏻‍♂️ 

I agree that the scope of vulns that this would avoid is very small, though, and it's not worth holding this up, so I changed it in [6f33e71](https://github.com/WordPress/two-factor/pull/487/commits/6f33e71f889c5618b6dfd59a795dc9414d84ccb0)